source/agora/consensus/SlashPolicy.d

```diff
@@ -0,0 +1,286 @@
+    - It determines what the penalty is for misbehaving validators -- in other
+        words, How many BOAs it will impose on the misbehaving validators
+        as a penalty
+
+    All the validators should reveal their pre-images timely in order for
+    the network to maintain randomness. So we need a penalty as an incentive
+    to make validators reveal them regularly. So this class manages all the
+    policies for penalty and slashing.
+    See https://github.com/bpfkorea/agora/issues/1076.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.consensus.SlashPolicy;
+
+import agora.common.Amount;
+import agora.common.crypto.Key;
+import agora.common.Types;
+import agora.consensus.EnrollmentManager;
+import agora.consensus.data.Params;
+import agora.consensus.data.Enrollment;
+import agora.consensus.data.Params;
+import agora.consensus.data.PreImageInfo;
+import agora.consensus.data.Transaction;
+import agora.consensus.state.UTXODB;
+import agora.utils.Log;
+version (unittest) import agora.utils.Test;
+
+import std.exception;
+
+mixin AddLogger!();
+
+/*******************************************************************************
+
+    Manage the policy for slashing
+
+*******************************************************************************/
+
+public class SlashPolicy
+{
+    // The amount of a penalty
+    public immutable Amount penalty_amount;
+
+    // The address to get a penalty
+    public immutable PublicKey penalty_address;
+
+    // Enrollment manager
+    private EnrollmentManager enroll_man;
+
+    /***************************************************************************
+
+        Constructor
+
+        Params:
+            enroll_man = the EnrollmentManager
+            params = the consensus-critical constants
+
+    ***************************************************************************/
+
+    public this (EnrollmentManager enroll_man, immutable(ConsensusParams) params)
+    {
+        this.enroll_man = enroll_man;
+        // TODO: This should be set through ConsensusParams
+        this.penalty_amount = Amount(10_000L * 10_000_000L);
+        this.penalty_address = params.CommonsBudgetAddress;
+    }
+
+    /***************************************************************************
+
+        Get the validators that do not reveal their pre-images and will be
+        slashed through the consensus procoess.
+
+        Params:
+            missing_validators = will contain the validators
+            height = the desired block height to look up the validators for
+
+    ***************************************************************************/
+
+    public void getMissingValidators (ref uint[] missing_validators,
+        in Height height) @safe nothrow
+    {
+        Hash[] keys;
```

Comment:
If you plan to use this as a buffer then I think you should add:

```
missing_validators.length = 0;
() @trusted { assumeSafeAppend(missing_validators); }();
```

Otherwise if you don't want it to be a buffer you can use `out uint[]`. But it would mean there is always an allocation. So the first approach might be better.